### PR TITLE
add key/value store service

### PR DESCRIPTION
**Pull request by DanCech**
This PR adds a simple key/value store, which we can use to store items that don't necessarily need their own separate table.

It currently uses the SQLStore as the backend, but can easily support other backends if that is desired.

### CODE REVIEW

**Comment by ryantxu** [comment]
@FZambia can you keep an eye on this issue :)

File: pkg/infra/kvstore/kvstore.go
Comment (ryantxu) [inline pkg/infra/kvstore/kvstore.go:42]: @DanCech thoughts on namespace constraints?   This may be similar to the live channel structure where we try to give a pattern that lets internal grafana, plugins, and datasources all use essentially named channels.  In the end, the channel path actually has orgId as a prefix see also https://github.com/grafana/grafana-plugin-sdk-go/blob/master/live/channel.go#L13

but I may be over thinking this -- for sure good to scope things in a way that plugins can safely use this same store

https://grafana.com/docs/grafana/latest/live/live-channel/#channel-structure

**Comment by FZambia** [comment]
Looks similar to remote cache implementations - maybe we should extend database cache to have non-expiring keys instead? What do you think?

**Comment by DanCech** [comment]
> Looks similar to remote cache implementations - maybe we should extend database cache to have non-expiring keys instead? What do you think?

That is a good point, and I did consider it but was worried about the difference in semantics.  The name "cache" implies to me that it's best-effort regardless of expiration (eg restarting a memory-only redis store will drop the cache) whereas this store is intended to be reliable.

File: pkg/infra/kvstore/kvstore.go
Comment (DanCech) [reply pkg/infra/kvstore/kvstore.go:42]: Interesting, definitely along the same lines.  I had envisioned the namespace basically combining the scope and namespace concepts you have in the channels, using namespaces like `plugin/<slug>` or `dashboard/<uid>` but hadn't formalized what that structure would look like.  We could adopt the same Scope concept here also if that makes sense to you.

If we're using a "normal" k/v store as a backend in the future we would very likely end up mapping to underlying keys like `<org>/<namespace>/<key>` (or `<org>/<scope>/<namespace>/<key>`)

Comment (FZambia) [reply pkg/infra/kvstore/kvstore.go:42]: > If we're using a "normal" k/v store as a backend in the future we would very likely end up mapping to underlying keys like <org>/<namespace>/<key> (or <org>/<scope>/<namespace>/<key>)

This exactly matches how we build channel names when passing over WebSocket to distinguish between the same channels in different orgs and in some in-memory data structures at the moment. I.e. `<org>/<scope>/<namespace>/<path>` - `key` is called `path` in Live case.

I don't see immediate benefits in attempt to combine two things into one concept, but have a feeling that maybe k/v storage should be more simplified here: i.e. just org_id and key fields (without namespace which can become a part of key itself) but enforce rules on a key creation so that all keys inherit `<scope>/<namespace>/<path>` pattern? At least this will help a bit to avoid possible key duplication in different parts of Grafana. But maybe over-engineering at this point since no clear benefit? 

Also there may be problems with MySQL when trying to build index like:

```
{Cols: []string{"org_id", "namespace", "key"}, Type: UniqueIndex},
```
Where namespace and key are both varchars 190 - there is a limit on index entry size as far as I remember so that MySQL will report errors like `Specified key was too long`

Comment (DanCech) [reply pkg/infra/kvstore/kvstore.go:42]: The key length issue is a good point, since we don't want to run into problems.  Anyone running mysql 5.7+ is unlikely to run into a problem unless they have explicitly disabled large indexes but it might be an issue for those on older mysql releases.  That said, there is already an index on `user (login, email)` which are both 190 char utf8mb4 columns (despite not being needed as there are also unique indexes on both `login` and `email`) so anyone with large indexes disabled would not be able to install Grafana AFAIK.

One of the reasons I had separated namespace from key was to be able to allocate 190 characters for the key (and another 190 for the namespace) rather than having a 190 character limit, it was also to make it easier to clean up entries later (eg to remove all entries for a specific namespace) without needing to resort to a LIKE query.

File: pkg/infra/kvstore/kvstore_test.go
Comment (bergquist) [inline pkg/infra/kvstore/kvstore_test.go:84]: In our 1:1 you mentioned that `0` could be global. But I don't see how we pick items with orgId=0 over orgId=1. But perhaps that convention is not needed yet and could just be ignored unless we actually need it.

**Review by bergquist — COMMENTED** [review]
Looks reasonable to me. Regarding the namespace/scope discussion I feel its hard to the right decision now unless we have a clear usage for this. For storing a few k/v items in the database this seems good enough IMO.

> or sure good to scope things in a way that plugins can safely use this same store

Offering storage for plugins would require more thought and aggressive quotas/limits to avoid malicious usage. 

@DanCech wanna mark this as ready? I guess @grafana/backend-platform should have their opinion as well.

File: pkg/infra/kvstore/kvstore.go
Comment (sdboyer) [inline pkg/infra/kvstore/kvstore.go:15]: Hard to put my finger on exactly why with my relatively minimal context, but something is tugging at me, saying "this ought to be an interface." Maybe because my first guess is that that clients ought not have to care about the difference between `KVStore` and `NamespacedKVStore`.

Comment (sdboyer) [reply pkg/infra/kvstore/kvstore.go:15]: Or maybe `KVStoreBackend` _is_ the abstracted interface clients are supposed to depend on. But in that case, why do we have these struct types wrapping it?

I'm just confused about this collection of interrelated types - i feel like i'm missing something about the intended use pattern, as what's here seems like more than what's minimally necessary.

File: pkg/infra/kvstore/kvstore.go
Comment (sdboyer) [inline pkg/infra/kvstore/kvstore.go:62]: Pseudoconstructors like this multiply the size of the state machine for the objects they're on - is it possible to refactor such that an `Init()` method isn't necessary?

**Review by sdboyer — COMMENTED** [review]
kv store seems to make sense in general; some particular questions about the design of the package and its types

File: pkg/infra/kvstore/kvstore.go
Comment (DanCech) [reply pkg/infra/kvstore/kvstore.go:15]: The intention here is for the consumer to interact only with `KVStore`, and the backend potentially be switched out depending on configuration.  So eg we could have an alternative backend based on redis.  With the upcoming switch to Wire we may be able to handle that in ProvideService and just have the backends implement an interface instead.

File: pkg/infra/kvstore/kvstore.go
Comment (DanCech) [reply pkg/infra/kvstore/kvstore.go:16]: I had structured things this way so that the dependency injection would work correctly, though there may be a cleaner way of doing that.

File: pkg/infra/kvstore/kvstore.go
Comment (DanCech) [reply pkg/infra/kvstore/kvstore.go:62]: If KVStore instantiated the logger and passed it to the backend then the backend wouldn't need to implement Init() which it currently uses to create a logger, I had structured it this way so that KVStore wouldn't need to deal with that.

File: pkg/infra/kvstore/kvstore.go
Comment (sdboyer) [reply pkg/infra/kvstore/kvstore.go:16]: Ahhh, the `inject`, got it

File: pkg/infra/kvstore/kvstore.go
Comment (sdboyer) [reply pkg/infra/kvstore/kvstore.go:15]: > The intention here is for the consumer to interact only with `KVStore`

But not `NamespacedKVStore`? My naive assumption would be that consumers would not know or care whether their k/v CRUD is namespaced, instead having that be a property of configuration mandated e.g. by dependency injection.

> With the upcoming switch to Wire we may be able to handle that in ProvideService and just have the backends implement an interface instead.

This really seems like what i was craving, both for this and the `Init()` comment below: a way of pushing this kind of initialization logic out to a choice made by the caller.

Comment (DanCech) [reply pkg/infra/kvstore/kvstore.go:15]: > But not `NamespacedKVStore`? My naive assumption would be that consumers would not know or care whether their k/v CRUD is namespaced, instead having that be a property of configuration mandated e.g. by dependency injection

There are 2 ways I was thinking about this:

1. Make it easier for something consuming the service to define a namespace in one place, eg:

```
	kvStore := uss.KVStore.WithNamespace(0, "my_namespace")
```

Then interact with `kvStore` to get/set keys rather than having to pass org id and namespace multiple times.

2. Provide a namespaced store object to components like plugins that should be able to get/set keys within their namespace but not outside of it.

> > With the upcoming switch to Wire we may be able to handle that in ProvideService and just have the backends implement an interface instead.
> 
> This really seems like what i was craving, both for this and the `Init()` comment below: a way of pushing this kind of initialization logic out to a choice made by the caller.

I'll take a look at how this could work with Wire and whether that gives us a pathway to simplify things.

File: pkg/infra/kvstore/kvstore_test.go
Comment (DanCech) [reply pkg/infra/kvstore/kvstore_test.go:84]: I haven't implemented anything like a fallback to a key in `orgId=0` if it's not found in a specific org, the intent was more that `orgId=0` could be used to store values that are not specific to a particular org.

File: pkg/infra/kvstore/kvstore.go
Comment (sdboyer) [reply pkg/infra/kvstore/kvstore.go:15]: > Provide a namespaced store object to components like plugins that should be able to get/set keys within their namespace but not outside of it.

Righto, this is the case i was naively imagining, with the main question being whether plugins ought to be aware that its k/v operations are namespaced. My intuition says no because that implies the existence of a clear decision rule for classifying a component as namespaceable-or-not, and it seems unlikely that such a rule exists.

> I'll take a look at how this could work with Wire and whether that gives us a pathway to simplify things.

Thank you! (@marefr you may have a take here)

**Review by marefr — APPROVED** [review]
LGTM

**Comment by FZambia** [comment]
Sent #38627 with possible tweaks, @DanCech plz take a look

**Comment by DanCech** [comment]
@sdboyer I think everything has been addressed now, can I get a final lgtm from you before we merge this?

File: pkg/server/wire.go
Comment (marefr) [inline pkg/server/wire.go:83]: Nit. ProvideService currently returns an interface. For every other "provider"/dependency we let the provider return the struct pointer and then use `wire.Bind` to bind to an interface instead.  I think we've sort of come to the conclusion that this is how we would like to manage things moving forward, basically _accept interfaces, return structs_.

**Review by marefr — APPROVED** [review]
LGTM, but added some additional comments.

File: pkg/server/wire.go
Comment (DanCech) [reply pkg/server/wire.go:83]: The intent in returning an interface is that ProvideService may in the future return an alternate implementation based on config, eg if we wanted to allow using a different storage backend.  I wanted to make it clear that the caller should not rely on that specific implementation but rather accept anything that implements the interface.  Maybe I'm getting ahead of myself there, what do you think?

Comment (marefr) [reply pkg/server/wire.go:83]: > I wanted to make it clear that the caller should not rely on that specific implementation

Agree with that. The caller will accept an interface and use that to interact with the store, but deciding what implementation to use will be handled by wire. At least with Enterprise this way allow Enterprise to depend on an OSS service (struct pointer), extending/overriding it with a different implementation. 

But yeah this not a big issue currently since kvstore it not used yet.

Comment (DanCech) [reply pkg/server/wire.go:83]: Ok, I'll merge it as-is and we can always tweak things later 👍

File: pkg/infra/kvstore/sql.go
Comment (DanCech) [reply pkg/infra/kvstore/sql.go:59]: I'm inclined not to worry about that, string comparison is still a very cheap operation relative to say database access.

File: pkg/infra/kvstore/kvstore.go
Comment (DanCech) [reply pkg/infra/kvstore/kvstore.go:10]: The intended use-case here would be to provide the option for an administrator to configure an alternative key/value backend for Grafana to use, not for a given Grafana instance to use multiple different backends.  An example use-case might be for Grafana Cloud if we decided to use a backend other than MySQL for scalability or operability reasons.

The namespace concept will likely be clearer when I post the PR I'm working on that uses the store, which depends on the KVStore service but only uses it to instantiate a NamespacedKVStore that it then interacts with (removing the need to pass the same org id and namespace in multiple places).  There may be other cases where we want to provide a NamespacedKVStore directly to a caller to restrict its access.

**Pull request merged**
